protocol/protocol.go

```diff
@@ -5,13 +5,615 @@ package protocol
+
+type RwritePkt struct {
+	RLen Count
+}
+
+type RerrorPkt struct {
+	Error string
+}
+
+type DirPkt struct {
+	D Dir
+}
+
+type RPCCall struct {
+	b     []byte
+	Reply chan []byte
+}
+
+type RPCReply struct {
+	b []byte
+}
+
+/* rpc servers */
+type ClientOpt func(*Client) error
+type ServerOpt func(*Server) error
+type Tracer func(string, ...interface{})
+
+// Client implements a 9p client. It has a chan containing all tags,
+// a scalar FID which is incremented to provide new FIDS (all FIDS for a given
+// client are unique), an array of MaxTag-2 RPC structs, a ReadWriteCloser
+// for IO, and two channels for a server goroutine: one down which RPCalls are
+// pushed and another from which RPCReplys return.
+// Once a client is marked Dead all further requests to it will fail.
+// The ToNet/FromNet are separate so we can use io.Pipe for testing.
+type Client struct {
+	Tags       chan Tag
+	FID        uint64
+	RPC        []*RPCCall
+	ToNet      io.WriteCloser
+	FromNet    io.ReadCloser
+	FromClient chan *RPCCall
+	FromServer chan *RPCReply
+	Msize      uint32
+	Dead       bool
+	Trace      Tracer
+}
+
+// Server is a 9p server.
+// For now it's extremely serial. But we will use a chan for replies to ensure that
+// we can go to a more concurrent one later.
+type Server struct {
+	NS        NineServer
+	D         Dispatcher
+	Versioned bool
+	FromNet   io.ReadCloser
+	ToNet     io.WriteCloser
+	Replies   chan RPCReply
+	Trace     Tracer
+	Dead      bool
+
+	fprofile *os.File
+}
+
+type NineServer interface {
+	Rversion(MaxSize, string) (MaxSize, string, error)
+	Rattach(FID, FID, string, string) (QID, error)
+	Rwalk(FID, FID, []string) ([]QID, error)
+	Ropen(FID, Mode) (QID, MaxSize, error)
+	Rcreate(FID, string, Perm, Mode) (QID, MaxSize, error)
+	Rstat(FID) ([]byte, error)
+	Rwstat(FID, []byte) error
+	Rclunk(FID) error
+	Rremove(FID) error
+	Rread(FID, Offset, Count) ([]byte, error)
+	Rwrite(FID, Offset, []byte) (Count, error)
+	Rflush(FID, FID) error
+}
 
-	"github.com/Harvey-OS/ninep/stub"
+var (
+	RPCNames = map[MType]string{
+		Tversion: "Tversion",
+		Rversion: "Rversion",
+		Tauth:    "Tauth",
+		Rauth:    "Rauth",
+		Tattach:  "Tattach",
+		Rattach:  "Rattach",
+		Terror:   "Terror",
+		Rerror:   "Rerror",
+		Tflush:   "Tflush",
+		Rflush:   "Rflush",
+		Twalk:    "Twalk",
+		Rwalk:    "Rwalk",
+		Topen:    "Topen",
+		Ropen:    "Ropen",
+		Tcreate:  "Tcreate",
+		Rcreate:  "Rcreate",
+		Tread:    "Tread",
+		Rread:    "Rread",
+		Twrite:   "Twrite",
+		Rwrite:   "Rwrite",
+		Tclunk:   "Tclunk",
+		Rclunk:   "Rclunk",
+		Tremove:  "Tremove",
+		Rremove:  "Rremove",
+		Tstat:    "Tstat",
+		Rstat:    "Rstat",
+		Twstat:   "Twstat",
+		Rwstat:   "Rwstat",
+	}
 )
 
-func NewServer(ns stub.NineServer, opts ...stub.ServerOpt) (*stub.Server, error) {
-	s := &stub.Server{}
-	s.Replies = make(chan stub.RPCReply, stub.NumTags)
+// GetTag gets a tag to be used to identify a message.
+func (c *Client) GetTag() Tag {
+	t := <-c.Tags
+	if false {
```

Comment:
we'll need to look at this to see whether to remove it at some point. I put the false here because I thought the finalizer was causing trouble; it was not. But if we remove the false we'll need to change the c.Tags <- *t below.